File: 42.Trapping Rain Water/TrappingRainWater.cpp
class Solution {
public:
    int trap(vector<int>& height) {
        int n=height.size();
        int left(0),water(0),level(0);//the height of water.
        int right=n-1;
        
        while(left<right)
        {
            if(height[left]<height[right])
            {
                level=max(height[left],level);
                water+=level-height[left++];
            }
            else{
                level=max(height[right],level);
                water+=level-height[right--];
            }
        }
        return water;
    }
};

File: 42.Trapping Rain Water/TrappingRainWater.java
public class Solution {
    public int trap(int[] height) {
        if(height == null || height.length == 0)
            return 0;
        int last = height.length - 1;
        int begin = 0;
        int level = Math.min(height[last], height[begin]);
        int water = 0;
        while( begin < last)
        {
            if(height[last] <= height[begin])
            {
                water += Math.max(0, level - height[last]);
                level = Math.max(height[last], level);
                last--;
            }
            else
            {
                water += Math.max(0, level - height[begin]);
                level = Math.max(height[begin], level);
                begin++;
            }
        }
        return water;
    }
}